addons/binding/org.openhab.binding.loxone/src/main/java/org/openhab/binding/loxone/internal/core/LxWsSecurity.java

```diff
@@ -0,0 +1,231 @@
+/**
+ * Copyright (c) 2010-2017 by the respective copyright holders.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.loxone.internal.core;
+
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+import java.util.function.BiConsumer;
+
+import javax.crypto.Mac;
+import javax.crypto.spec.SecretKeySpec;
+
+import org.apache.commons.codec.DecoderException;
+import org.apache.commons.codec.binary.Hex;
+import org.openhab.binding.loxone.internal.core.LxJsonResponse.LxJsonSubResponse;
+import org.openhab.binding.loxone.internal.core.LxServer.Configuration;
+import org.openhab.binding.loxone.internal.core.LxWsClient.LxWebSocket;
+
+/**
+ * Security abstract class providing authentication and encryption services.
+ * Used by the {@link LxWsClient} during connection establishment to authenticate user and during message exchange for
+ * encryption and decryption or the messages.
+ *
+ * @author Pawel Pieczul - initial contribution
+ *
+ */
+abstract class LxWsSecurity {
+    int debugId;
+    String user;
+    String password;
+    Thread executor;
+    LxWebSocket socket;
+    Configuration configuration;
+    LxOfflineReason reason;
+    String details;
+    boolean cancel = false;
+
+    /**
+     * Create an authentication instance.
+     *
+     * @param debugId
+     *            instance of the client used for debugging purposes only
+     * @param configuration
+     *            configuration object for getting and setting custom properties (e.g. token)
+     * @param socket
+     *            websocket to perform communication with Miniserver
+     * @param user
+     *            user to authenticate
+     * @param password
+     *            password to authenticate
+     */
+    LxWsSecurity(int debugId, Configuration configuration, LxWebSocket socket, String user, String password) {
+        this.debugId = debugId;
```

Comment:
Make these fields `protected final` (or `private` if not accessed outside this class)?

Reply:
final sure, but protected seems to be larger than no modifier at all, they are accessed by the subclasses but of the same package